### PR TITLE
DAOS-5654 IV: refine pool_iv_conn pic_size handling

### PR DESCRIPTION
**Pull request by liuxuezhao**
Refine a few details of valid checking pool_iv_conn.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

File: src/pool/srv_iv.c [hunk 1]
```diff
@@ -302,7 +302,7 @@ pool_iv_conn_lookup(struct pool_iv_conns *conns, uuid_t uuid)
 	struct pool_iv_conn	*conn = conns->pic_conns;
 	char			*end = (char *)conn + conns->pic_size;
 
-	while ((char *)conn < end) {
+	while (conn->pic_creds <= end) {
```
Comment (wangdi1) [inline src/pool/srv_iv.c:305]: conn + pool_iv_conn_size(conn->pic_cred_size) ?

File: src/pool/srv_iv.c [hunk 2]
```diff
@@ -327,11 +326,12 @@ pool_iv_conn_delete(struct pool_iv_conns *conns, uuid_t hdl_uuid)
 		pool_iv_conn_size(conn->pic_cred_size));
 
 	size = pool_iv_conn_size(conn->pic_cred_size);
+	D_ASSERT(conns->pic_size >= size);
 	end = (char *)conns->pic_conns + conns->pic_size;
-	next = (char *)pool_iv_conn_next(conn);
-	if (end != next)
+	next = pool_iv_conn_next(conn);
+	if (next->pic_creds <= end)
```
Comment (wangdi1) [inline src/pool/srv_iv.c:332]: same?

File: src/pool/srv_iv.c [hunk 3]
```diff
@@ -491,12 +492,12 @@ pool_iv_ent_copy(struct ds_iv_key *key, d_sg_list_t *dst, d_sg_list_t *src)
 
 			conn = src_conns->pic_conns;
 			end = (char *)conn + src_conns->pic_size;
-			while ((char *)conn < end) {
+			while (conn->pic_creds <= end) {
```
Comment (wangdi1) [inline src/pool/srv_iv.c:495]: same

File: src/pool/srv_iv.c [hunk 4]
```diff
@@ -730,7 +731,7 @@ pool_iv_ent_refresh(struct ds_iv_entry *entry, struct ds_iv_key *key,
 
 		conn = src_iv->piv_conn_hdls.pic_conns;
 		end = (char *)conn + src_iv->piv_conn_hdls.pic_size;
-		while ((char *)conn < end) {
+		while (conn->pic_creds <= end) {
```
Comment (wangdi1) [inline src/pool/srv_iv.c:734]: same

**Review by wangdi1 — APPROVED** [review]
Actually DAOS-5654 should be resolved by PR-3425.  The real reason is that iv_buffer for fetching is not initialized.

File: src/pool/srv_iv.c (hunk 1 above)
Comment (liw) [inline src/pool/srv_iv.c:305]: What is the problem with the original code?

File: src/pool/srv_iv.c (hunk 2 above)
Comment (liw) [inline src/pool/srv_iv.c:332]: Similarly, what the problem with the original code?

**Review by liw — COMMENTED** [review]
Sorry, a couple of questions in line.

File: src/pool/srv_iv.c (hunk 1 above)
Comment (liuxuezhao) [reply src/pool/srv_iv.c]: done

Comment (liuxuezhao) [reply src/pool/srv_iv.c]: In my test I ever got a segfault in pool_iv_conn_lookup() as described in DAOS-5654 (not reproducible every time).
By checking code, it seems to me that  original checking conn valid by "conn->pic_creds <= end" looks not enough, for example a few garbage or dirty bytes left in previous pool_iv_conns but no enough space for one whole iv_conn?

**Review by liw — APPROVED** [review]

File: src/pool/srv_iv.c (hunk 1 above)
Comment (liw) [reply src/pool/srv_iv.c]: It feels like the real issue is that conns isn't properly set up. If that issue is fixed, the original checks here should work fine.

Comment (liuxuezhao) [reply src/pool/srv_iv.c]: i suspected it too, but did not find other errors when insert/delete.
on the other hand, if the pic_cred_size possible difference for different pool conn, it seems possible leave garbage data when different pools connect/disconnect, this is what this patch want to refine.

Comment (liw) [reply src/pool/srv_iv.c]: OK... Thanks for elaborating.

File: src/pool/srv_iv.c [hunk 5]
```diff
@@ -296,13 +296,20 @@ pool_iv_conn_ent_size(size_t cred_size)
 	       pool_iv_conn_size(cred_size);
 }
 
+static bool
+pool_iv_conn_valid(struct pool_iv_conn *conn, char *end)
+{
+	return (conn->pic_creds <= end) &&
+	       ((char *)conn + pool_iv_conn_size(conn->pic_cred_size) <= end);
```
Comment (wangdi1) [inline src/pool/srv_iv.c:303]: only conn + pool_iv_conn_size() is enough here?

**Review by wangdi1 — APPROVED** [review]

File: src/pool/srv_iv.c (hunk 5 above)
Comment (liuxuezhao) [reply src/pool/srv_iv.c:303]: I was think that when "conn->pic_creds <= end" not true, need not calculate pool_iv_conn_size(conn->pic_cred_size) as that is possibly incorrect already.

**Pull request merged**